[MRG] More informative error message in OneHotEncoder(categories=None) with negative integer values

ygivenx opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #12179 - Update the error message when negative inputs are provided to OneHotEncoder to show how to negative inputs can be supported by instantiating OneHotEncoder with categories='auto'
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
OneHotEncoder didn't support negative integer inputs. This behavior is 
deprecated. The legacy behaviour currently raises ValueError("X needs 
to contain only non-negative integers."). This error message is now 
extended to show how to activate the new behavior by instantiating the 
OneHotEncoder with categories='auto'.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Fix Issue #12179

OneHotEncoder "only non-negative integers" message should suggest using
categories='auto'

Commit:
Fix Issue #12179

    OneHotEncoder "only non-negative integers" message should suggest using
    categories='auto'

ramanujam commented:
Fix Issue #12179
    
    OneHotEncoder "only non-negative integers" message should suggest using
    categories='auto'

jnothman commented:
Please change the title of the pull request to be more descriptive.

ramanujam commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -421,7 +421,10 @@ def _legacy_fit_transform(self, X):
         X = check_array(X, dtype=np.int)
         if np.any(X < 0):
-            raise ValueError("X needs to contain only non-negative integers.")
+            raise ValueError("X needs to contain only non-negative integers,"
```
Sure, thanks

ramanujam replied:
Fixed.

Commit:
Fix Issue #12179

OneHotEncoder "only non-negative integers" message should suggest using
categories='auto'

ogrisel commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -421,7 +421,8 @@ def _legacy_fit_transform(self, X):
         if np.any(X < 0):
-            raise ValueError("X needs to contain only non-negative integers.")
+            raise ValueError("Instantiate using categories='auto' to"
+                             "handle negative integer categories.")
```
May I suggest the following error message that I find more explicit:

OneHotEncoder in legacy mode cannot handle categories encoded as negative integers. Please set categories='auto' explicitly to be able to use arbitrary integer values as category identifiers.

ramanujam replied:
Sure, Will do.

ogrisel commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -504,7 +505,8 @@ def _legacy_transform(self, X):
         if np.any(X < 0):
-            raise ValueError("X needs to contain only non-negative integers.")
+            raise ValueError("Instantiate using categories='auto' to"
+                             "handle negative integer categories.")
```
Same remark here.

ogrisel reviewed (commented):
Can you also please add or update a test to check that is the exception with the correct message is raised in the expected cases (both for fit and transform)?

ogrisel commented:
I don't think it's worth the effort, there is already a FutureWarning and this code will be scrapped in 2 releases.

Commit:
Fixes #12180 Modify the error message

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -421,7 +421,11 @@ def _legacy_fit_transform(self, X):
         X = check_array(X, dtype=np.int)
         if np.any(X < 0):
-            raise ValueError("X needs to contain only non-negative integers.")
+            raise ValueError("OneHotEncoder in legacy mode cannot handle"
```
Missing space between "handle" and "categories" and similar on subsequent lines

jorisvandenbossche reviewed (approved):
Apart from the spacing-comment of @jnothman , looks good to me

Commit:
Fix the spacing
